# SELF STUDY for "PANDAS FOR DATA ANALYSIS" with FERUAZABONU KHURRAMOVA #

### I am going to do this assignment with a real-world, high-volume dataset: the NBA Games and Player Statistics (1946–2015) dataset. This dataset is perfect for testing advanced filtering, grouping, and complex merging logic. ###

### 🏀Step 1: Professional Setup & DataFrames (Section 1) ###

### Part A: Task 1.1 — Creating Manual Metadata ##

In [7]:
import pandas as pd

# Task 1.1: Recreating the structure with professional sports data
df_teams = pd.DataFrame({
    "team_id": ["BOS", "LAL", "CHI"],
    "titles": [17, 17, 6],
    "city": ["Boston", "Los Angeles", "Chicago"]
})

# Code answers:
print(f"Type of one column: {type(df_teams['titles'])}") # Returns Series
print(f"Type of selection: {type(df_teams[['team_id', 'titles']])}") # Returns DataFrame
print(f"Default Index: {df_teams.index}") # Returns RangeIndex

Type of one column: <class 'pandas.Series'>
Type of selection: <class 'pandas.DataFrame'>
Default Index: RangeIndex(start=0, stop=3, step=1)


### Part B: Task 1.2 — Expanding the Metadata ###

Task 1.2

In [8]:
# Task 1.2: Adding rows and resetting index
extra_teams = pd.DataFrame({
    "team_id": ["GSW", "SAS", "PHI"],
    "titles": [7, 5, 3],
    "city": ["San Francisco", "San Antonio", "Philadelphia"]
})

df_teams = pd.concat([df_teams, extra_teams]).reset_index(drop=True)
df_teams

,team_id,titles,city
0,BOS,17,Boston
1,LAL,17,Los Angeles
2,CHI,6,Chicago
3,GSW,7,San Francisco
4,SAS,5,San Antonio
5,PHI,3,Philadelphia


### 🏀 Step 2: Loading & Professional Exploration (Section 2) ###

### Part A: Task 2.1 — Smart Loading ###

In [9]:
import pandas as pd

# Task 2.1: Loading from the local file in your explorer
# We use the exact name seen in your sidebar
try:
    df = pd.read_csv("nbaallelo.csv.csv", parse_dates=["date_game"])
    print("✅ Success! Local NBA Dataset loaded.")
    display(df.head(3))
except FileNotFoundError:
    # If the first name fails, try the alternative
    df = pd.read_csv("nbaallelo.csv", parse_dates=["date_game"])
    print("✅ Success! NBA Dataset loaded using alternative name.")
    display(df.head(3))

✅ Success! Local NBA Dataset loaded.


,gameorder,game_id,lg_id,_iscopy,year_id,date_game,seasongame,is_playoffs,team_id,fran_id,...,win_equiv,opp_id,opp_fran,opp_pts,opp_elo_i,opp_elo_n,game_location,game_result,forecast,notes
0,1,194611010TRH,NBA,0,1947,1946-11-01,1,0,TRH,Huskies,...,40.294830,NYK,Knicks,68,1300.0000,1306.7233,H,L,0.640065,NaN
1,1,194611010TRH,NBA,1,1947,1946-11-01,1,0,NYK,Knicks,...,41.705170,TRH,Huskies,66,1300.0000,1293.2767,A,W,0.359935,NaN
2,2,194611020CHS,NBA,0,1947,1946-11-02,1,0,CHS,Stags,...,42.012257,NYK,Knicks,47,1306.7233,1297.0712,H,W,0.631101,NaN


### Part B: Task 2.2 — Professional Inspection ###

In [10]:
# 1. Check the scale of the data
print(f"Dataset Shape: {df.shape}")

# 2. Check column types and missing values
df.info()

# 3. See statistical summaries (like average points)
display(df.describe())

# 4. Count missing values per column
print(df.isna().sum())

Dataset Shape: (126314, 23)
<class 'pandas.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   gameorder      126314 non-null  int64         
 1   game_id        126314 non-null  str           
 2   lg_id          126314 non-null  str           
 3   _iscopy        126314 non-null  int64         
 4   year_id        126314 non-null  int64         
 5   date_game      126314 non-null  datetime64[us]
 6   seasongame     126314 non-null  int64         
 7   is_playoffs    126314 non-null  int64         
 8   team_id        126314 non-null  str           
 9   fran_id        126314 non-null  str           
 10  pts            126314 non-null  int64         
 11  elo_i          126314 non-null  float64       
 12  elo_n          126314 non-null  float64       
 13  win_equiv      126314 non-null  float64       
 14  opp_id         126314 non-null  str

,gameorder,_iscopy,year_id,date_game,seasongame,is_playoffs,pts,elo_i,elo_n,win_equiv,opp_pts,opp_elo_i,opp_elo_n,forecast
count,126314.000000,126314.000000,126314.000000,126314,126314.000000,126314.000000,126314.000000,126314.000000,126314.000000,126314.000000,126314.000000,126314.000000,126314.000000,126314.000000
mean,31579.000000,0.500000,1988.200374,1988-04-09 00:53:18.429311,43.533733,0.063857,102.729982,1495.236055,1495.236055,41.707889,102.729982,1495.236055,1495.236055,0.500000
min,1.000000,0.000000,1947.000000,1946-11-01 00:00:00,1.000000,0.000000,0.000000,1091.644500,1085.774400,10.152501,0.000000,1091.644500,1085.774400,0.020447
25%,15790.000000,0.000000,1975.000000,1974-11-08 00:00:00,22.000000,0.000000,93.000000,1417.237975,1416.994900,34.103035,93.000000,1417.237975,1416.994900,0.327989
50%,31579.000000,0.500000,1990.000000,1990-02-06 00:00:00,43.000000,0.000000,103.000000,1500.945550,1500.954400,42.113357,103.000000,1500.945550,1500.954400,0.500000
75%,47368.000000,1.000000,2003.000000,2003-03-19 00:00:00,65.000000,0.000000,112.000000,1576.060000,1576.291625,49.635328,112.000000,1576.060000,1576.291625,0.672011
max,63157.000000,1.000000,2015.000000,2015-06-16 00:00:00,108.000000,1.000000,186.000000,1853.104500,1853.104500,71.112038,186.000000,1853.104500,1853.104500,0.979553
std,18231.927643,0.500002,17.582309,NaN,25.375178,0.244499,14.814845,112.139945,112.461687,10.627332,14.814845,112.139945,112.461687,0.215252


gameorder             0
game_id               0
lg_id                 0
_iscopy               0
year_id               0
date_game             0
seasongame            0
is_playoffs           0
team_id               0
fran_id               0
pts                   0
elo_i                 0
elo_n                 0
win_equiv             0
opp_id                0
opp_fran              0
opp_pts               0
opp_elo_i             0
opp_elo_n             0
game_location         0
game_result           0
forecast              0
notes            120890
dtype: int64


### Task 3.1 — Advanced Filtering ###

In [11]:
# Task 3.1: Filtering by Multiple Conditions
# We want high-scoring modern games
modern_high_score = df[(df["pts"] > 150) & (df["year_id"] > 2010)]

print(f"Found {len(modern_high_score)} modern games with > 150 points.")
display(modern_high_score[["date_game", "team_id", "pts", "opp_id", "opp_pts"]])

Found 0 modern games with > 150 points.


,date_game,team_id,pts,opp_id,opp_pts


### Task 3.2 — The Power of GroupBy ###

In [12]:
# Task 3.2: Grouping and Aggregating
# Calculate average points per franchise and sort them
top_scoring_teams = df.groupby("fran_id")["pts"].mean().sort_values(ascending=False)

print("Top 5 Highest Scoring Franchises (Historical Average):")
print(top_scoring_teams.head(5))

Top 5 Highest Scoring Franchises (Historical Average):
fran_id
Condors       115.446512
Squires       114.051314
Floridians    112.654545
Stars         112.354497
Colonels      111.625296
Name: pts, dtype: float64


### Section 4: Data Cleaning (Final Step for Part A) ###

In [13]:
# Task 4.1: Handling Missing Values
# Since 'notes' is mostly empty, we will fill missing ones with 'no notes'
df["notes"] = df["notes"].fillna("no notes")

# Verify that there are now 0 missing values
print("Remaining missing values in 'notes':", df["notes"].isna().sum())

Remaining missing values in 'notes': 0


##### 1. Data Interpretation ##### 
Based on my analysis:

The franchise with the highest historical scoring average is the Condors (~115.45 pts).

I successfully filled 120,890 missing values in the notes column with the string "no notes" to ensure the dataset is complete for future processing.

##### 2. Quick Check Answers #####
Why use & instead of and? We use & because it is a bitwise operator that can compare entire Pandas Series element-by-element. The standard and only works on single True/False values.

Why are parentheses required? Parentheses are mandatory around each condition because bitwise operators like & have higher precedence than comparison operators (like >). Without them, Python gets confused about the order of operations.

##### 3. Self-Study Reflection #####
I have completed the core tasks of Module 2, Class 4. I practiced:

Section 1: Creating and modifying DataFrames manually.

Section 2: Loading a large local CSV (nbaallelo.csv.csv) and performing data health inspections.

Section 3: Using complex boolean filters and groupby aggregations.

Section 4: Handling missing values (fillna).

The Final Missing Piece: Task 3.4 — Merging

In [14]:
# Task 3.4: Merging with a Lookup Table
# 1. Create a small lookup DataFrame manually
# This maps the Franchise ID to their Home City and Arena Capacity
lookup_df = pd.DataFrame({
    "fran_id": ["Condors", "Squires", "Floridians", "Stars"],
    "home_city": ["Pittsburgh", "Virginia", "Miami", "Utah"],
    "arena_capacity": [12500, 10000, 15000, 19000]
})

# 2. Merge it into your main NBA DataFrame
# We use a 'left' join to keep all our NBA games
df_final = pd.merge(df, lookup_df, on="fran_id", how="left")

# 3. Check the new columns at the end
print("New columns added via merge:")
display(df_final[["date_game", "fran_id", "home_city", "arena_capacity"]].head())

New columns added via merge:


,date_game,fran_id,home_city,arena_capacity
0,1946-11-01,Huskies,NaN,NaN
1,1946-11-01,Knicks,NaN,NaN
2,1946-11-02,Stags,NaN,NaN
3,1946-11-02,Knicks,NaN,NaN
4,1946-11-02,Falcons,NaN,NaN


What happens to rows with no match? Since we used a left join, all original NBA rows are kept. For franchises not in our small lookup_df (like the Lakers or Celtics), the new home_city and arena_capacity columns will simply contain NaN (Not a Number) values.